File: packages/telemetry/typescript/src/sdk/tracer.ts
import { type Attributes, type Context, type Span, type SpanOptions, type Tracer, trace } from "@opentelemetry/api"
import { ATTRIBUTES } from "../constants/attributes.ts"
import { SCOPE_LATITUDE } from "../constants/scope.ts"
import type { ContextOptions } from "./types.ts"

/**
 * Returns an OpenTelemetry tracer scoped under Latitude's instrumentation namespace.
 * Spans created with this tracer pass the LatitudeSpanProcessor's smart filter.
 */
export function getLatitudeTracer(scope: string): Tracer {
  return trace.getTracer(`${SCOPE_LATITUDE}.${scope}`)
}

/**
 * Builds the Latitude span attributes for a context. Mirrors the keys the
 * LatitudeSpanProcessor stamps from the active context, so spans carrying these
 * attributes directly are indistinguishable from capture()-scoped spans on ingest.
 */
export function latitudeAttributesFromContext(options: ContextOptions): Attributes {
  const attributes: Attributes = {}
  const project = options.project ?? options.projectSlug

  if (options.tags && options.tags.length > 0) attributes[ATTRIBUTES.tags] = JSON.stringify(options.tags)
  if (options.metadata && Object.keys(options.metadata).length > 0) {
    attributes[ATTRIBUTES.metadata] = JSON.stringify(options.metadata)
  }
  if (options.sessionId) attributes[ATTRIBUTES.sessionId] = options.sessionId
  if (options.userId) attributes[ATTRIBUTES.userId] = options.userId
  if (options.userEmail) attributes[ATTRIBUTES.userEmail] = options.userEmail
  if (options.memoryScope) attributes[ATTRIBUTES.memoryScope] = options.memoryScope
  if (project) attributes[ATTRIBUTES.project] = project

  return attributes
}

/**
 * Wraps a tracer so every span it starts carries the given attributes. Lets a framework
 * that owns the model call (e.g. Cloudflare Think via `experimental_telemetry.tracer`)
 * attach per-turn Latitude context without an ambient context — which matters on runtimes
 * whose AsyncLocalStorage lacks `enterWith()`, such as Cloudflare Workers.
 */
export function withLatitudeAttributes(tracer: Tracer, attributes: Attributes): Tracer {
  if (Object.keys(attributes).length === 0) return tracer

  const stamp = (span: Span): Span => {
    span.setAttributes(attributes)
    return span
  }

  // The inner `startActiveSpan` is generic over the callback type; delegating through a
  // loosely-typed reference lets us splice in the attribute-stamping wrapper without the
  // overload/generic friction.
  const startActive = tracer.startActiveSpan.bind(tracer) as (name: string, ...rest: unknown[]) => unknown

  const wrapper: Tracer = {
    startSpan(name: string, options?: SpanOptions, context?: Context): Span {
      return stamp(tracer.startSpan(name, options, context))
    },
    startActiveSpan<F extends (span: Span) => unknown>(
      name: string,
      arg1: F | SpanOptions,
      arg2?: F | Context,
      arg3?: F,
    ): ReturnType<F> {
      const wrap = (fn: F) => (span: Span) => fn(stamp(span))
      if (typeof arg1 === "function") {
        return startActive(name, wrap(arg1)) as ReturnType<F>
      }
      if (typeof arg2 === "function") {
        return startActive(name, arg1, wrap(arg2)) as ReturnType<F>
      }
      return startActive(name, arg1, arg2 as Context, wrap(arg3 as F)) as ReturnType<F>
    },
  }

  return wrapper
}

File: packages/telemetry/python/src/latitude_telemetry/sdk/tracer.py
import json
from collections.abc import Generator, Sequence
from contextlib import contextmanager
from typing import Any, cast

from opentelemetry.context import Context
from opentelemetry.trace import Link, Span, SpanKind, Tracer
from opentelemetry.util.types import Attributes, AttributeValue

from latitude_telemetry.constants import ATTRIBUTES, SCOPE_LATITUDE
from latitude_telemetry.sdk._deprecation import warn_project_slug_deprecated
from latitude_telemetry.sdk.types import ContextOptions


def _scope_name(scope: str) -> str:
    return scope if scope == SCOPE_LATITUDE or scope.startswith(f"{SCOPE_LATITUDE}.") else f"{SCOPE_LATITUDE}.{scope}"


LatitudeAttributes = dict[str, AttributeValue]


def latitude_attributes_from_context(options: ContextOptions) -> LatitudeAttributes:
    attributes: LatitudeAttributes = {}
    project = options.get("project")
    if project is None and "project_slug" in options:
        warn_project_slug_deprecated("capture")
        project = options.get("project_slug")

    tags = options.get("tags")
    metadata = options.get("metadata")
    session_id = options.get("session_id")
    user_id = options.get("user_id")
    user_email = options.get("user_email")
    if tags:
        attributes[ATTRIBUTES.tags] = json.dumps(tags)
    if metadata:
        attributes[ATTRIBUTES.metadata] = json.dumps(metadata)
    if session_id:
        attributes[ATTRIBUTES.session_id] = session_id
    if user_id:
        attributes[ATTRIBUTES.user_id] = user_id
    if user_email:
        attributes[ATTRIBUTES.user_email] = user_email
    if project:
        attributes[ATTRIBUTES.project] = project

    return attributes


class _LatitudeTracer:
    def __init__(self, tracer: Tracer, attributes: LatitudeAttributes):
        self._tracer = tracer
        self._attributes = attributes

    def start_span(
        self,
        name: str,
        context: Context | None = None,
        kind: SpanKind = SpanKind.INTERNAL,
        attributes: Attributes | None = None,
        links: Sequence[Link] | None = None,
        start_time: int | None = None,
        record_exception: bool = True,
        set_status_on_exception: bool = True,
    ) -> Span:
        span = self._tracer.start_span(
            name,
            context=context,
            kind=kind,
            attributes=attributes,
            links=links,
            start_time=start_time,
            record_exception=record_exception,
            set_status_on_exception=set_status_on_exception,
        )
        span.set_attributes(self._attributes)
        return span

    @contextmanager
    def start_as_current_span(
        self,
        name: str,
        context: Context | None = None,
        kind: SpanKind = SpanKind.INTERNAL,
        attributes: Attributes | None = None,
        links: Sequence[Link] | None = None,
        start_time: int | None = None,
        record_exception: bool = True,
        set_status_on_exception: bool = True,
        end_on_exit: bool = True,
    ) -> Generator[Span, None, None]:
        with self._tracer.start_as_current_span(
            name,
            context=context,
            kind=kind,
            attributes=attributes,
            links=links,
            start_time=start_time,
            record_exception=record_exception,
            set_status_on_exception=set_status_on_exception,
            end_on_exit=end_on_exit,
        ) as span:
            span.set_attributes(self._attributes)
            yield span

    def __getattr__(self, name: str) -> Any:
        return getattr(self._tracer, name)


def with_latitude_attributes(tracer: Tracer, attributes: LatitudeAttributes) -> Tracer:
    if not attributes:
        return tracer
    return cast(Tracer, _LatitudeTracer(tracer, attributes))


def get_latitude_tracer(provider: Any, scope: str, context: ContextOptions | None = None) -> Tracer:
    tracer = provider.get_tracer(_scope_name(scope))
    if context is None:
        return tracer
    return with_latitude_attributes(tracer, latitude_attributes_from_context(context))
